File: testbed_ruby/test/associations/many_to_n_test.rb
require 'test/unit'
require './src-gen-umple/student_q'
require './src-gen-umple/mentor_q'
require './src-gen-umple/program_q'

module CruiseTest
class ManyToNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentQ.new(99)
    assert_equal(0,s.number_of_mentors)
  end
  
  def test_ConstructorWithTooFew
    assert_raise RuntimeError do #"Unable to create MentorQ, must have 2 students"
      s = StudentQ.new(99)
      MentorQ.new("blah",[s])
    end
  end
  
  def test_ConstructorWithTooMany
    assert_raise RuntimeError do #"Unable to create MentorQ, must have 2 students"
      s = StudentQ.new(99)
      s2 = StudentQ.new(98)    
      s3 = StudentQ.new(97)
      MentorQ.new("blah",[s,s2,s3])
    end
  end
  
  def test_constructorJustRight
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)
  
    m = MentorQ.new("blah",[s,s2])
  
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
  end
  
  def test_constructorWatchOutForDuplicateEntries
    assert_raise RuntimeError do #"Unable to create MentorQ, must have 2 students"
      s = StudentQ.new(99)
      MentorQ.new("blah",[s, s])
    end
  end
  
  def test_constructorExistingMentorsOkay
    s1 = StudentQ.new(99)
    s2 = StudentQ.new(98)
    s3 = StudentQ.new(97)
    
    m = MentorQ.new("blah",[s1,s2])
  
    assert_equal(2,m.number_of_students)
    assert_equal(m,s1.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)    
  
    m2 = MentorQ.new("blah2",[s2,s3])
    
    assert_equal(2,m.number_of_students)
    assert_equal(m,s1.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    
    assert_equal(2,m2.number_of_students)
    assert_equal(m2,s2.get_mentor(1))
    assert_equal(m2,s3.get_mentor(0))
  end
  
  def test_set_students_outsideN
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)    
    s3 = StudentQ.new(97)
    s4 = StudentQ.new(96)
    s5 = StudentQ.new(95)
  
    m = MentorQ.new("blah",[s,s2])
    
    assert_equal(false,m.set_students([s2,s3,s4]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  
    assert_equal(false,m.set_students([]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)
  end
  
  def test_set_students_doNotAllowDuplicatesNM
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)    
    s3 = StudentQ.new(97)
  
    m = MentorQ.new("blah",[s,s2])
    assert_equal(false,m.set_students([s2,s2,s3]))
  end
  
  def test_set_students_withinN
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)    
    s3 = StudentQ.new(97)
  
    m = MentorQ.new("blah",[s,s2])
    
    assert_equal(true,m.set_students([s3,s2]))
    assert_equal(2,m.number_of_students)
  
    assert_equal(0,s.number_of_mentors)
  
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))
  end
  
  def test_set_students_empty
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)
    m = MentorQ.new("blah",[s,s2])
  
    assert_equal(false,m.set_students([]))
    assert_equal(2,m.number_of_students)
  end
  
  def test_deletStudentDeletesTheMentorIfNotEnoughLeft
    s1 = StudentQ.new(99)
    s2 = StudentQ.new(98)
  
    m = MentorQ.new("blah",[s1,s2])
    studentP = ProgramQ.new
    mentorP = ProgramQ.new
    
    s2.set_program(studentP)
    m.set_program(mentorP)
  
    s1.delete
    
    
    assert_equal(0,m.number_of_students)
    assert_equal(nil,m.get_program)
    assert_equal(nil,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s1.number_of_mentors)
  end    
  
  def test_add_mentorTooMany
    s = StudentQ.new(99)
    s2 = StudentQ.new(98)    
    s3 = StudentQ.new(97)
  
    m = MentorQ.new("blah",[s,s2])
    
    assert_equal(false,s3.add_mentor(m))
    assert_equal(2,m.number_of_students)
    assert_equal(0,s3.number_of_mentors)
  end
end
end

File: testbed_php/test/associations/ManyToNTest.php
<?php

class ManyToNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentQ(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  
  public function test_ConstructorWithTooFew()
  {
  	$this->expectException(new Exception("Unable to create MentorQ, must have 2 students"));
    $s = new StudentQ(99);
    new MentorQ("blah",array($s));
  }

  public function test_ConstructorWithTooMany()
  {
  	$this->expectException(new Exception("Unable to create MentorQ, must have 2 students"));
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);    
    $s3 = new StudentQ(97);
    new MentorQ("blah",array($s,$s2,$s3));
  }
  
  

  public function test_constructorJustRight()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);

    $m = new MentorQ("blah",array($s,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
  }
  
  public function test_constructorWatchOutForDuplicateEntries()
  {
  	$this->expectException(new Exception("Unable to create MentorQ, must have 2 students"));
    $s = new StudentQ(99);
    new MentorQ("blah",array($s, $s));
  }
  

  public function test_constructorExistingMentorsOkay()
  {
    $s1 = new StudentQ(99);
    $s2 = new StudentQ(98);
    $s3 = new StudentQ(97);
    
    $m = new MentorQ("blah",array($s1,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s1->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual(0,$s3->numberOfMentors());    

    $m2 = new MentorQ("blah2",array($s2,$s3));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s1->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual($m2,$s2->getMentor_index(1));
    $this->assertEqual($m2,$s3->getMentor_index(0));

  }   
  

  public function test_setStudents_outsideN()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);    
    $s3 = new StudentQ(97);
    $s4 = new StudentQ(96);
    $s5 = new StudentQ(95);

    $m = new MentorQ("blah",array($s,$s2));
    
    $this->assertEqual(false,$m->setStudents($s2,$s3,$s4));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());

    $this->assertEqual(false,$m->setStudents($s5));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor_index(0));
    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());
    
    
  }


  public function test_setStudents_doNotAllowDuplicatesNM()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);    
    $s3 = new StudentQ(97);

    $m = new MentorQ("blah",array($s,$s2));
    $this->assertEqual(false,$m->setStudents($s2,$s2,$s3));
  }


  public function test_setStudents_withinN()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);    
    $s3 = new StudentQ(97);

    $m = new MentorQ("blah",array($s,$s2));
    
    $this->assertEqual(true,$m->setStudents(array($s3,$s2)));
    $this->assertEqual(2,$m->numberOfStudents());

    $this->assertEqual(0,$s->numberOfMentors());

    $this->assertEqual($m,$s2->getMentor_index(0));
    $this->assertEqual($m,$s3->getMentor_index(0));
  }
  



  public function test_setStudents_empty()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);
    $m = new MentorQ("blah",array($s,$s2));

    $this->assertEqual(false,$m->setStudents(array()));
    $this->assertEqual(2,$m->numberOfStudents());
  }
  

  public function test_deletStudentDeletesTheMentorIfNotEnoughLeft()
  {
    $s1 = new StudentQ(99);
    $s2 = new StudentQ(98);

    $m = new MentorQ("blah",array($s1,$s2));
    $studentP = new ProgramQ();
    $mentorP = new ProgramQ();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(null,$m->getProgram());
    $this->assertEqual(null,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s1->numberOfMentors());
  }  
    

  public function test_addMentorTooMany()
  {
    $s = new StudentQ(99);
    $s2 = new StudentQ(98);    
    $s3 = new StudentQ(97);

    $m = new MentorQ("blah",array($s,$s2));
    
    $this->assertEqual(false,$s3->addMentor($m));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(0,$s3->numberOfMentors());
    
  }

}

?>
